Do not wrap function calls using stream in mirror

MnlPhlp opened the pull request:
## Changes

When using a mirrored stuct as type in a StreamSink, do not wrap the function call in mirror_<struct>() in the generated rust code. Fixes #1275.

At the moment there is the problem that the mirrored types have to manually implement intoDart for the StreamSink to work. I have no good idea how we could change that because we cant implement a trait on structs from another crate.
Any ideas on how to fix this? I just created this pull-request at this point to discuss this problem. Obviously this needs to be fixed before merging

I added a test that is failing at this point. If there is no good way around manually implementing intoDart i can do that for a test. 

## Checklist

- [x] An issue to be fixed by this PR is listed above.
- [x] New tests are added to ensure new features are working. End-to-end tests are usually in the `./frb_example/pure_dart` example, more specifically, `rust/src/api.rs` and `dart/lib/main.dart`.
- [ ] The code generator is run and the code is formatted (via `just precommit`).
- [ ] If this PR adds/changes features, documentations (in the `./book` folder) are updated.
- [ ] CI is passing.

fzyzcjy commented:
> At the moment there is the problem that the mirrored types have to manually implement intoDart for the StreamSink to work. I have no good idea how we could change that because we cant implement a trait on structs from another crate.

Hmm good question. So first idea may be, why this is not a problem for all other mirror types? Maybe we can just mimic it.

MnlPhlp commented:
We would need StreamSink<mirror_<struct>> as type. That should probably work with some adjustments in the generator

MnlPhlp commented:
using StreamSink<mirror_Struct> works now. What do you think about it?
I don't think the failing ci jobs are because of this right?

fzyzcjy commented:
Nice!

> I don't think the failing ci jobs are because of this right?

Seems unrelated. For MacOS, it is a long standing problem. For web, maybe create an issue at related repos?

MnlPhlp commented:
I changed the logic a bit.
Instead of looking for the mirror_ prefix in a bunch of places i remove it when parsing.
Seems to work fine and vectors of mirrored types now also work as Stream argument

MnlPhlp commented:
Added another tests for mirror_ structs in structs defined in the api. Inspired from #1282 .
I also added some notes to the chapter in the book.

I think that's it for this PR if you don't notice any problems.

MnlPhlp commented:
I discovered an error when generating without any existing generated files. Tests didn't show that. Maybe the tests could clear all generated files before running. 
But i have no time to look into this more

I fixed the error, it is just not covered by the tests

fzyzcjy commented on book/src/feature/stream.md:
```diff
@@ -19,3 +19,36 @@ See [logging examples](logging.md) which uses streams extensively.
+
+## Streaming external types (mirror)
+
+When using mirrored external types in a StreamSink you have to specify the generated `mirror_..` type in the StreamSink. This is because the generated code can not implement the necessary interfaces on a struct from a different crate. The generated `mirror_struct` implements `From<struct>` so you can use your normal types and just call `.into()` before passing them to the StreamSink.
```
Ok I start reviewing this PR. Good job! Though I am a bit hesitate about this developer interface. Is it possible we do not leak the "mirror_something" types to end users? That looks like internal implementation details, and we should hide them.

MnlPhlp replied:
The problem is that Stream sink requires the inner type to implement IntoDart and we can not implement that on external types.
Maybe we could require the type for the stream to implement Into<> for some type that implements IntoDart. But I am not sure that is possible.
We could also hide the mirror type with a macro that converts from struct to mirror_struct but I don't think that is much better.

Something like StreamSink<mirror!(ExternalStruct)>

fzyzcjy replied:
Yes I agree that's not trivial to solve.

For brainstom, some more naive thoughts (may not be good): We may implement StreamSink to convert the types internally (from ExternalStruct to mirror_ExternalStruct).

But we do not want to change signature visible to the user (e.g. `StreamSink<ExternalStruct, mirror_ExternalStruct>` does not sound very beautiful.

Then maybe

```rust
// static code
trait StreamSink<T> { ... }

// generated code
// since it is generated, we are free to generate the code that converts ExternalStruct into mirror_ExternalStruct, just like what we did for return values
struct _StreamSinkExternalStruct;
impl StreamSink<ExternalStruct> for _StreamSinkExternalStruct {...}

// again generated code
fn generated_bridge_function_that_calls_f() {
  f(_StreamSinkExternalStruct(...))
}

// user in api.rs
fn f(stream: impl StreamSink<ExternalStruct>) {}
```

fzyzcjy replied:
Also cc @Desdaemon in case you want to play with these :)

MnlPhlp replied:
I got something like that working (not generated yet, just typed out)

generated code:
```rust
pub trait LogEntryStreamSink {
    fn add(&self, val: LogEntry);
}
impl LogEntryStreamSink for StreamSink<LogEntry> {
    fn add(&self, val: LogEntry) {
        self.add_inner::<_, LogEntry>(val);
    }
}
```

api
```rust
use crate::bridge_generated::VecBleDeviceStreamSink;
pub fn create_log_stream(s: StreamSink<LogEntry>) {
    s.add(LogEntry {})
}

```

You still have to import a trait to get the method but we don't have the mirror_ type in the api anymore.
`add_inner` was previously the add method on a sink. I had to change some type constraints for this to work.

If this interface looks good i will try to generate the code automatically. I had to add some type annotations but only in the generated.rs file.

MnlPhlp replied:
That way of doing it would also make all other changes from this PR obsolete :D
I tried it on a new branch without any of the other changes I made so far.

fzyzcjy replied:
You proposal looks also interesting!

A bit of worry: What if the user wants to use StreamSink generically?
But that should not be a big problem, because the user can just call `add_inner` (and manually pass the type argument).

> I tried it on a new branch without any of the other changes I made so far.

Btw you can make a "draft PR" for that branch, b/c it will be easy to see `Files changed` with a (draft) PR

MnlPhlp commented:
Replaced by #1285
